### PR TITLE
enable clippy lints against integer casts

### PR DESCRIPTION
**Pull request by RalfJung**
Cc https://github.com/rust-lang/miri/issues/1236

### CODE REVIEW

File: src/shims/mod.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#![warn(clippy::integer_arithmetic)]
```
Comment (oli-obk) [inline src/shims/mod.rs:1]: Not sure how useful this one is... none of he cases are caught, so we could just as well turn on debug asserts in release mode if we haven't already. Would be more readable than the checked math functions

Comment (RalfJung) [reply src/shims/mod.rs:1]: > turn on debug asserts in release mode

Any suggestion for how to do that in a way that also applies to the builds produced by rustc? We cannot use cargo profiles.

Also we do have a bunch of arithmetic outside this folder (stacked borrows, data races, weak memory) that I deliberately did not enable this lint for, so there might be a performance difference. But that should be easy to benchmark.

Comment (oli-obk) [reply src/shims/mod.rs:1]: We can change Bootstrap's build flags for miri

Comment (oli-obk) [reply src/shims/mod.rs:1]: Yea I don't expect this to make a difference. It's mostly in math heavy code where this is visible in practice

Comment (RalfJung) [reply src/shims/mod.rs:1]: > We can change Bootstrap's build flags for miri

How? RUSTFLAGS affects all dependencies so it'll rebuild everything...

I am going to move the contentious part into a separate PR.

**Comment by RalfJung** [comment]
@oli-obk I hope the remaining changes are uncontroversial. :)

**Comment by oli-obk** [comment]
They are

@bors r+

**Comment by RalfJung** [comment]
@bors retry

I fixed the PR title

**Pull request merged**